allow visualization postprocessors to use the current cell

jdannberg opened the pull request:
This pull request adds an active cell iterator (instead of a reference to the active cell iterator, which was used before, and is now marked as deprecated) to the material model inputs structure, which allows the cell to be filled in the constructor of material model inputs that is used for the visualization postprocessors (using `get_cell<DoFHandler<dim>`). 

This is useful in cases where the material model uses the cell information to extract, for example, the old solution, or some additional solution variables that are not in the material model inputs, like in the case of melt transport, and the computes the material model outputs using this information. With this new version, visualization postprocessors will now also be able to output these quantities from material model outputs correctly, which was not possible before.

jdannberg commented:
I had to update some of the tests, but the only one where I had to update the output was the cell_reference test, which now outputs the cell reference in more places, so I think that makes sense. But @gassmoeller, I think you wrote the test, maybe you could have a look?

Also, I am not sure why the copy constructor of MaterialModelInputs works if it is called as `MaterialModelInputs<dim> in (base_inputs)`, but not when we call `MaterialModelInputs<dim> in(n_points, n_comp); in = base_inputs` as in the Newton solver test cases. So it would be great if I could get some help with that, as I think the old version should also continue to work (as for now I just changed how we call the constructor in those test files).

bangerth commented on source/material_model/interface.cc:
```diff
@@ -236,7 +236,7 @@ namespace aspect
 
 
-
+    DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
```
can you say why this is necessary?

jdannberg replied:
Otherwise we get a warning that we use cell, which is deprecated (and I can not put the block just around the single line, because it's in the constructor). 
Or did you mean I should add a comment why this is necessary?

bangerth replied:
comments never hurt, but I primarily wanted to understand it from the perspective of a reviewer :-)

bangerth commented on source/material_model/interface.cc:
```diff
@@ -288,7 +290,6 @@ namespace aspect
-
     template <int dim>
     MaterialModelInputs<dim>::MaterialModelInputs(const FEValuesBase<dim,dim> &fe_values,
                                                   const typename DoFHandler<dim>::active_cell_iterator *cell_x,
```
It would of course be nice to also make this into a reference rather than a pointer. how many places would this affect?

In the interest of keeping review simple, maybe keep that for a follow-up patch if you want to do it. Or open an issue about it.

jdannberg replied:
I opened an issue for now (#1907) so that we don't forget about it, and I can do it after this pull request is merged.

bangerth commented on source/material_model/interface.cc:
```diff
@@ -303,12 +304,28 @@ namespace aspect
       strain_rate(fe_values.n_quadrature_points, numbers::signaling_nan<SymmetricTensor<2,dim> >()),
-      cell(cell_x)
+      cell(cell_x),
+      current_cell(*cell_x)
```
What happens if `cell_x` is a `NULL` pointer? Don't you have to do something like
```
  current_cell (cell_x == NULL ? typename DoFHandler<dim>::active_cell_iterator() : *cell_x)
```
?

I think it would probably be good to also do
```
  cell (cell_x == NULL ? NULL : &current_cell)
```
This way, the calling site definitely no longer needs to keep the pointed to object around, and this avoids a whole class of possible errors.

bangerth commented on source/material_model/interface.cc:
```diff
@@ -342,7 +359,11 @@ namespace aspect
       this->cell = cell_x;
+      DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
+
+      this->current_cell = *cell_x;
```
same as above, you'll have to be careful if `cell_x==NULL`

bangerth commented on source/postprocess/visualization/melt.cc:
```diff
@@ -104,9 +104,9 @@ namespace aspect
+        // Set use_strain_rates to true since the compaction viscosity might also depend on the strain rate.
         MaterialModel::MaterialModelInputs<dim> in(input_data,
-                                                   this->introspection(), false);
+                                                   this->introspection(), true);
```
This change is independent, right? Can you break it into a separate PR?

bangerth commented:
What was the error message you got with the old-style
```
MaterialModelInputs<dim> in(n_points, n_comp); 
in = base_inputs;
```
call? I don't think the class has a copy constructor so you may have to initialize everything during construction time now. But the error message would tell :-)

jdannberg commented:
I fixed it with help from @gassmoeller. I had to add an assignment operator to MaterialModelInputs, because with the default one, the value that was copied for the cell was not a valid iterator state, and that was not allowed.

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -217,6 +217,16 @@ namespace aspect
+      /**
+       * Assignment operator.
+       */
+      MaterialModelInputs &operator=(const MaterialModelInputs &material);
```
I'm not actually a fan of this. These objects are big, and copying them is expensive. I'd rather just outlaw it (as it was before, implicitly by just not providing this constructor).

jdannberg replied:
Of course I can just revert this. But I was just thinking of @MFraters. Those tests where the assignment was used were all related to the Newton solver, and I was thinking he might have a number of his own plugins and test cases that might use a similar logic.

bangerth replied:
Hm, but then I'm curious again how this worked before. If you remove the operator, what are the errors you get from the existing code?

I'm not saying that you should revert the patch, I'm just curious to see what changed that required you to add the copy operator. I must be missing something that it's not clear to me why you need it.

gassmoeller replied:
Removing this operator and the constructor would not help, because they are implicitly created when called, and even worse you then get a cryptic error message that you can not assign a DoFAccessor without a particle handler (apparently `active_cell_iterator a = active_cell_iterator()` is not allowed). That is why we created the constructor and operator. Should we instead mark them as deleted in the header file like this?
```
MaterialModelInputs (...) = delete;
... operator= (...) = delete;
```

bangerth replied:
Or are you saying that this leads to an error:
```
  active_cell_iterator cell;  // invalid state
  active_cell_iterator cell2;
  cell2 = cell;    // copy of invalid state
```

bangerth replied:
Ah, wires crossed. The `=delete` approach doesn't work because that's C++11. We need to come up with something else.

jdannberg replied:
The error message is:
`An error occurred in line <115> of file </home/dannberg/software/dealii-8.5-install/include/deal.II/dofs/dof_accessor.templates.h> in function
    void dealii::DoFAccessor<structdim, DoFHandlerType, lda>::set_dof_handler(DoFHandlerType*) [with int structdim = 2; DoFHandlerType = dealii::DoFHandler<2>; bool level_dof_access = false]
The violated condition was: 
    dh != 0
Additional information: 
    (none)`

Basically, I think it tries to set a dof_handler, but does not have one.

jdannberg replied:
> 
> 
> Or are you saying that this leads to an error:
> ```
>   active_cell_iterator cell;  // invalid state
>   active_cell_iterator cell2;
>   cell2 = cell;    // copy of invalid state
> ```
> 

Yes, exactly.

MFraters replied:
@jdannberg, I am fine with changing it to the `MaterialModelInputs<dim> in (base_inputs)` style in the Newton tests. I indeed have some more tests which are not part of the repository  yet, but changing it should not be much work.

bangerth replied:
Thanks, @MFraters .

@jdannberg -- what an annoying issue. I'm fixing this in deal.II, but this is of course too late for your current patch.

jdannberg commented:
@bangerth, I think I addressed all of your comments except for the issue with the assignment operator. As @MFraters said it would be fine to change the constructor to the `MaterialModelInputs<dim> in (base_inputs)` style, I'd be totally fine with that, in this case we just have to figure out what to do about the cryptic error message people will get if they try to use the assignment operator anywhere without having the cell object. Do you have a good idea that works without marking the assignment operator as deleted? Should I just change the implementation to throw an assert that says you can not assign a material model inputs object without having a cell reference? Or is there a better way?

tjhei commented:
> Or is there a better way?

You can also declare it but make it private and don't provide an implementation. That is the pre-c++11 way of deleting it.

tjhei commented on source/material_model/interface.cc:
```diff
@@ -342,7 +385,12 @@ namespace aspect
+      DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
+
+      if (cell_x != NULL)
+        this->current_cell = *cell_x;
```
don't you have to reset ``current_cell`` if ``cell_x==NULL``?

jdannberg replied:
Oh, yes, I didn't think about that case. Thanks! Should be fixed now.

bangerth replied:
Oh, great catch. I missed that one as well!

tjhei commented on tests/cell_reference/screen-output:
```diff
@@ -34,6 +34,14 @@ Level: 1 Index: 2
 Level: 1 Index: 2
 Level: 1 Index: 3
 Level: 1 Index: 3
+Level: 1 Index: 0
```
what happened here?

jdannberg replied:
We now have the cell object in more places than before (because one of the constructors of MaterialModelInputs was changed to use the get_cell function). So I thought it made sense that there is more output now (the evaluate function outputs the cell if it gets it).

jdannberg commented:
Okay, thanks, sounds good to me!
@bangerth what do you think, would that be a good solution?

bangerth commented:
Yes, making it `private` is the correct approach. This leads to an error message that you may want to explain in the documentation of the function, namely that copying such objects is forbidden because it is too expensive.

bangerth commented:
I've gone through it one more time and it looks good.

Please move the copy operator into the `private` section of the class, squash the commits into one, and merge yourself. Thanks for your patience!
